### PR TITLE
Update JNI Target relocation record to have offset to the reloLocation.

### PR DESCRIPTION
**Pull request by dsouzai**
On x86, the JNI Target relocation puts the incorrect address when
generating a Runtime Assumption. However, the address that the
relocation infra itself needs to patch is correct. Thus, the relo record
has to hold not just the location to patch but also the location to
register the assumption against. This is the same on most platforms, but
different on x86.

~Depends on https://github.com/eclipse/omr/pull/6326~
~Depends on https://github.com/eclipse/omr/pull/6331~
Depends on https://github.com/eclipse/omr/pull/6332

Fixes https://github.com/eclipse-openj9/openj9/issues/14300

### CODE REVIEW

**Comment by dsouzai** [comment]
@jdmpapin do you mind reviewing this and the associated OMR PR?

**Review by mstoodle — APPROVED** [review]
couple of minor comments, but looks ok to me if the 8bit mask is fixed/explained :)

File: runtime/compiler/codegen/J9AheadOfTimeCompile.cpp [hunk 1]
```diff
@@ -345,6 +345,28 @@ J9::AheadOfTimeCompile::initializeCommonAOTRelocationHeader(TR::IteratedExternal
       case TR_JNIVirtualTargetAddress:
       case TR_JNIStaticTargetAddress:
       case TR_JNISpecialTargetAddress:
+         {
+         TR_RelocationRecordDirectJNICall *djnicRecord = reinterpret_cast<TR_RelocationRecordDirectJNICall *>(reloRecord);
+         TR_RelocationRecordInformation *recordInfo = reinterpret_cast<TR_RelocationRecordInformation*>(relocation->getTargetAddress());
+
+         uintptr_t offsetToReloLocation = recordInfo->data1;
+         TR_ASSERT_FATAL((offsetToReloLocation & 0xFF) == 0,
```
[the remark this answers is not in the document]
Comment (dsouzai) [reply runtime/compiler/codegen/J9AheadOfTimeCompile.cpp:353]: Ah right, yeah; I wanted to check only the least significant bits can be non zero.

[the remark this answers is not in the document]
Comment (dsouzai) [reply runtime/compiler/codegen/J9AheadOfTimeCompile.cpp]: Thanks, addressed in force push (https://github.com/eclipse-openj9/openj9/compare/231e5fd0bfbd1c712de8ed121495ba82e16f7f68..e355c30d8681c81d1548c6793860e5ee4e739d8e)

File: runtime/compiler/runtime/RelocationRecord.cpp [hunk 2]
```diff
@@ -1940,52 +1945,63 @@ TR_RelocationRecordDirectJNIVirtualMethodCall::getMethodFromCP(TR_RelocationRunt
...
 
-   createJNICallSite((void *)ramMethod, (void *)reloLocation,getMetadataAssumptionList(reloRuntime->exceptionTable()));
+   createJNICallSite((void *)ramMethod, addressForAssumption, getMetadataAssumptionList(reloRuntime->exceptionTable()));
    RELO_LOG(reloRuntime->reloLogger(), 6, "\t\tapplyRelocation: registered JNI Call redefinition site\n");
 
    reloTarget->storeRelativeAddressSequence((uint8_t *)newAddress, reloLocation, fixedSequence1);
    return 0;
+   }
+
+int32_t TR_RelocationRecordRamMethodConst::applyRelocation(TR_RelocationRuntime *reloRuntime, TR_RelocationTarget *reloTarget, uint8_t *reloLocation)
```
[the remark this answers is not in the document]
Comment (dsouzai) [reply runtime/compiler/runtime/RelocationRecord.cpp]: Addressed in force push (https://github.com/eclipse-openj9/openj9/compare/231e5fd0bfbd1c712de8ed121495ba82e16f7f68..e355c30d8681c81d1548c6793860e5ee4e739d8e)

File: runtime/compiler/codegen/J9AheadOfTimeCompile.cpp [hunk 3]
```diff
@@ -345,6 +345,28 @@ J9::AheadOfTimeCompile::initializeCommonAOTRelocationHeader(TR::IteratedExternal
...
+         {
+         TR_RelocationRecordDirectJNICall *djnicRecord = reinterpret_cast<TR_RelocationRecordDirectJNICall *>(reloRecord);
+         TR_RelocationRecordInformation *recordInfo = reinterpret_cast<TR_RelocationRecordInformation*>(relocation->getTargetAddress());
+
+         uintptr_t offsetToReloLocation = recordInfo->data1;
+         TR_ASSERT_FATAL((offsetToReloLocation & ~0xFF) == 0,
+                         "offsetToReloLocation %" OMR_PRIuPTR " cannot fit in a uint8_t",
+                         offsetToReloLocation);
+
+         TR::SymbolReference symRef* = reinterpret_cast<TR::SymbolReference *>(recordInfo->data2);
```
[the remark this answers is not in the document]
Comment (dsouzai) [reply runtime/compiler/codegen/J9AheadOfTimeCompile.cpp]: Fixed in force push.

File: runtime/compiler/runtime/RelocationRecord.cpp [hunk 4]
```diff
@@ -168,6 +168,11 @@ struct TR_RelocationRecordJ2IVirtualThunkPointerBinaryTemplate : public TR_Reloc
    IDATA _offsetToJ2IVirtualThunkPointer;
    };
 
+struct TR_RelocationRecordDirectToJNIBinaryTemplate : public TR_RelocationRecordConstantPoolBinaryTemplate
```
[the remark this answers is not in the document]
Comment (dsouzai) [reply runtime/compiler/runtime/RelocationRecord.cpp]: Yeah good catch; fixed in force push.

File: runtime/compiler/runtime/RelocationRecord.cpp [hunk 5]
```diff
@@ -1940,52 +1945,64 @@ TR_RelocationRecordDirectJNIVirtualMethodCall::getMethodFromCP(TR_RelocationRunt
...
    {
-   TR_RelocationRuntimeLogger *reloLogger = reloRuntime->reloLogger();
-   J9ConstantPool * newConstantPool =(J9ConstantPool *) computeNewConstantPool(reloRuntime, reloTarget, constantPool(reloTarget));
-   TR_OpaqueMethodBlock *ramMethod = getMethodFromCP(reloRuntime, newConstantPool, cpIndex(reloTarget));
-
-   if (!ramMethod)
-      {
-      return compilationAotClassReloFailure;
-      }
+   reloTarget->storeUnsigned8b(offsetToReloLocation, (uintptr_t *) &((TR_RelocationRecordDirectToJNIBinaryTemplate *)_record)->_offsetToReloLocation);
```
[the remark this answers is not in the document]
Comment (dsouzai) [reply runtime/compiler/runtime/RelocationRecord.cpp]: Fixed in force push.

File: runtime/compiler/p/codegen/J9TreeEvaluator.cpp [hunk 6]
```diff
@@ -12022,11 +12022,20 @@ TR::Instruction *loadAddressJNI32(TR::CodeGenerator *cg, TR::Node * node, int32_
...
+         int16_t inlinedSiteIndex = node ? node->getInlinedSiteIndex() : -1;
+         info->data3 = static_cast<uintptr_t>(inlinedSiteIndex);
+
+         cg->addExternalRelocation(
+            new (cg->trHeapMemory()) TR::BeforeBinaryEncodingExternalRelocation(
+               cursor,
+               reinterpret_cast<uint8_t *>(info),
+               reloType,
+               cg),
+            __FILE__, __LINE__, node);
```
[the remark this answers is not in the document]
Comment (dsouzai) [reply runtime/compiler/p/codegen/J9TreeEvaluator.cpp]: Fixed in force push.

File: runtime/compiler/runtime/RelocationRecord.cpp [hunk 7]
```diff
@@ -1940,52 +1945,64 @@ TR_RelocationRecordDirectJNIVirtualMethodCall::getMethodFromCP(TR_RelocationRunt
...
    {
-   TR_RelocationRuntimeLogger *reloLogger = reloRuntime->reloLogger();
-   J9ConstantPool * newConstantPool =(J9ConstantPool *) computeNewConstantPool(reloRuntime, reloTarget, constantPool(reloTarget));
-   TR_OpaqueMethodBlock *ramMethod = getMethodFromCP(reloRuntime, newConstantPool, cpIndex(reloTarget));
-
-   if (!ramMethod)
-      {
-      return compilationAotClassReloFailure;
-      }
+   reloTarget->storeUnsigned8b(offsetToReloLocation, (uint8_t *) &((TR_RelocationRecordDirectToJNIBinaryTemplate *)_record)->_offsetToReloLocation);
```
Comment (jdmpapin) [inline runtime/compiler/runtime/RelocationRecord.cpp:1951]: This doesn't necessarily need to change, but `&(...)->_offsetToReloLocation` is already a `uint8_t*`

**Review by jdmpapin — APPROVED** [review]

File: runtime/compiler/runtime/RelocationRecord.cpp (hunk 7 above)
Comment (dsouzai) [reply runtime/compiler/runtime/RelocationRecord.cpp]: Fixed in latest force push.

**Comment by dsouzai** [comment]
Gonna have to open yet another PR; because I force pushed, it's causing OpenJ9 OMR mirroring to fail.

**Comment by jdmpapin** [comment]
Jenkins test sanity+aot all jdk17 depends eclipse/omr#6332

**Comment by dsouzai** [comment]
Worth noting all tests passted except for aarch64, which failed because of
```
16:13:25  /home/jenkins/workspace/Build_JDK17_aarch64_linux_Personal/openj9/runtime/compiler/aarch64/codegen/ARM64JNILinkage.cpp: In member function 'TR::Instruction* J9::ARM64::JNILinkage::generateMethodDispatch(TR::Node*, bool, TR::RegisterDependencyConditions*, uintptr_t, TR::Register*)':
16:13:25  /home/jenkins/workspace/Build_JDK17_aarch64_linux_Personal/openj9/runtime/compiler/aarch64/codegen/ARM64JNILinkage.cpp:924:51: error: invalid use of member function 'TR::Compilation* OMR::Linkage::comp()' (did you forget the '()' ?)
16:13:25         TR_RelocationRecordInformation *info = new (comp->trHeapMemory()) TR_RelocationRecordInformation();
16:13:25                                                     ^~~~
16:13:25  /home/jenkins/workspace/Build_JDK17_aarch64_linux_Personal/openj9/runtime/compiler/aarch64/codegen/ARM64JNILinkage.cpp:924:55: error: base operand of '->' is not a pointer
16:13:25         TR_RelocationRecordInformation *info = new (comp->trHeapMemory()) TR_RelocationRecordInformation();
16:13:25                                                         ^~
```
Will force the build break fix. Commenting this so that we don't need to run tests on all platforms again.

**Comment by dsouzai** [comment]
Jenkins test sanity+aot all jdk17 depends https://github.com/eclipse/omr/pull/6332

**Comment by dsouzai** [comment]
Ok well jenkins is too unstable to launch the PR tests so I guess we're on standby till then..

**Comment by pshipton** [comment]
Jenkins test sanity.functional+aot all jdk17 depends https://github.com/eclipse/omr/pull/6332

**Comment by AdamBrousseau** [comment]
Jenkins test sanity.functional+aot all jdk17 depends eclipse/omr#6332

**Comment by dsouzai** [comment]
The failing tests are due to 
1. https://github.com/eclipse-openj9/openj9/issues/14461
2. https://github.com/eclipse-openj9/openj9/issues/14459

1 should be fixed now and the test in 2 has been disabled.

**Comment by dsouzai** [comment]
Jenkins test sanity.functional+aot all jdk17 depends eclipse/omr#6332

launching again just for sanity purposes.

**Comment by dsouzai** [comment]
The other aix failure is in `cmdLineTester_libpathTestRtf_0` caused by
```
[2022-02-10T01:50:23.505Z]  [ERR] Exception in thread "main" java.awt.AWTError: Can't connect to X11 window server using 'unix:0' as the value of the DISPLAY variable.
[2022-02-10T01:50:23.505Z]  [ERR] 	at java.desktop/sun.awt.X11GraphicsEnvironment.initDisplay(Native Method)
[2022-02-10T01:50:23.505Z]  [ERR] 	at java.desktop/sun.awt.X11GraphicsEnvironment$1.run(X11GraphicsEnvironment.java:101)
[2022-02-10T01:50:23.505Z]  [ERR] 	at java.base/java.security.AccessController.doPrivileged(AccessController.java:683)
[2022-02-10T01:50:23.505Z]  [ERR] 	at java.desktop/sun.awt.X11GraphicsEnvironment.<clinit>(X11GraphicsEnvironment.java:60)
[2022-02-10T01:50:23.505Z]  [ERR] 	at java.desktop/sun.awt.PlatformGraphicsInfo.createGE(PlatformGraphicsInfo.java:36)
[2022-02-10T01:50:23.505Z]  [ERR] 	at java.desktop/java.awt.GraphicsEnvironment$LocalGE.createGE(GraphicsEnvironment.java:93)
[2022-02-10T01:50:23.505Z]  [ERR] 	at java.desktop/java.awt.GraphicsEnvironment$LocalGE.<clinit>(GraphicsEnvironment.java:84)
[2022-02-10T01:50:23.505Z]  [ERR] 	at java.desktop/java.awt.GraphicsEnvironment.getLocalGraphicsEnvironment(GraphicsEnvironment.java:106)
[2022-02-10T01:50:23.505Z]  [ERR] 	at java.desktop/sun.awt.X11.XToolkit.<clinit>(XToolkit.java:224)
[2022-02-10T01:50:23.505Z]  [ERR] 	at java.desktop/sun.awt.PlatformGraphicsInfo.createToolkit(PlatformGraphicsInfo.java:40)
[2022-02-10T01:50:23.505Z]  [ERR] 	at java.desktop/java.awt.Toolkit.getDefaultToolkit(Toolkit.java:599)
[2022-02-10T01:50:23.505Z]  [ERR] 	at java.desktop/java.awt.Toolkit.getEventQueue(Toolkit.java:1493)
[2022-02-10T01:50:23.505Z]  [ERR] 	at java.desktop/java.awt.EventQueue.isDispatchThread(EventQueue.java:1087)
[2022-02-10T01:50:23.505Z]  [ERR] 	at java.desktop/javax.swing.SwingUtilities.isEventDispatchThread(SwingUtilities.java:1493)
[2022-02-10T01:50:23.505Z]  [ERR] 	at java.desktop/javax.swing.text.StyleContext.reclaim(StyleContext.java:473)
[2022-02-10T01:50:23.505Z]  [ERR] 	at java.desktop/javax.swing.text.StyleContext.addAttribute(StyleContext.java:330)
[2022-02-10T01:50:23.505Z]  [ERR] 	at java.desktop/javax.swing.text.StyleContext$NamedStyle.addAttribute(StyleContext.java:1558)
[2022-02-10T01:50:23.505Z]  [ERR] 	at java.desktop/javax.swing.text.StyleContext$NamedStyle.setName(StyleContext.java:1368)
[2022-02-10T01:50:23.505Z]  [ERR] 	at java.desktop/javax.swing.text.StyleContext$NamedStyle.<init>(StyleContext.java:1315)
[2022-02-10T01:50:23.505Z]  [ERR] 	at java.desktop/javax.swing.text.StyleContext.addStyle(StyleContext.java:125)
[2022-02-10T01:50:23.505Z]  [ERR] 	at java.desktop/javax.swing.text.StyleContext.<init>(StyleContext.java:105)
[2022-02-10T01:50:23.505Z]  [ERR] 	at java.desktop/javax.swing.text.DefaultStyledDocument.<init>(DefaultStyledDocument.java:109)
[2022-02-10T01:50:23.505Z]  [ERR] 	at org.openj9.test.libpath.Rtf.convert(Rtf.java:42)
[2022-02-10T01:50:23.505Z]  [ERR] 	at org.openj9.test.libpath.Rtf.main(Rtf.java:60)
```
which happens on the second run. Considering the method at the top of the stack is a native method, maybe there's something wrong with the relocation on power.

**Comment by dsouzai** [comment]
DirectToJNI is disabled on P as the query [1] is not overridden. All of my changes are related to the following methods:
```
loadAddressRAM32
loadAddressRAM
loadAddressJNI32
loadAddressJNI
```
These only run when generating directToJNI code.

I verified that the relocation isn't generated by running a unit test @jdmpapin has written. As such, the failing test can't be related to my changes.

[1] https://github.com/eclipse/omr/blob/e6d7abfdac20e3c3e3325afbc855fbe7342ea5c2/compiler/codegen/OMRCodeGenerator.hpp#L1514

**Comment by dsouzai** [comment]
@jdmpapin do you mind merging this if everything looks good to you?

**Comment by jdmpapin** [comment]
All observed test failures are unrelated just as @dsouzai described above. Re-launching tests for a coordinated merge (both to include jdk8,jdk11 and to ensure that the test results will be recent)

Jenkins test sanity all jdk8,jdk11,jdk17 depends eclipse/omr#6332

**Comment by pshipton** [comment]
Should we be adding this fix to the Java 18 release?

**Comment by dsouzai** [comment]
We could. It's a very low likelihood issue in practice (which explains why it's only been reported now) but at the same time, it's 100% reproducible in Devin's unit test. What kind of changes are generally expected to be cherry-picked into a release? If it's *any* kind of bug fix then I guess this would fall into that category.

**Comment by pshipton** [comment]
We want to deliver fixes that may impact users and are of low risk for breaking something else. If it's low risk we could just put it in, even if not sure of the impact on users.

**Comment by pshipton** [comment]
Sometimes we want changes to stew for a while in the head stream before backporting, to ensure they aren't going to cause additional issues. If this is one of those things then given "very low likelihood in practice" we shouldn't backport if it needs to stew longer than it already has.

**Comment by dsouzai** [comment]
Ok I'll open a PR for openj9-omr/openj9 0.31 shortly.

**Comment by dsouzai** [comment]
> Sometimes we want changes to stew for a while in the head stream before backporting, to ensure they aren't going to cause additional issues. If this is one of those things then given "very low likelihood in practice" we shouldn't backport if it needs to stew longer than it already has.

Sorry, I didn't seet his when I posted my earlier comment (and I just got the email notification for it for some reason). 

I mean the PR involves changing the raw relocation data stored to the SCC for these directToJNI relocations, which had to be done across all platforms, so while the change is relatively straightforward, it does have a big scope. Additionally there is a workaround for the problem, namely `-Xaot:disableDirectToJNIInline`.

The combination of these two facts leads me to personally believe we don't _need_ to double deliver. However, if others feel differently I have no problem opening the necessary PRs. Thoughts @mstoodle @DanHeidinga ?

Also I just realized I should have also changed the JITserver minor version -_-

**Comment by mstoodle** [comment]
Agree, we haven't seen it in practice until now so I think it can wait for 0.32 . Can/has Devin's test case be(en) added to our tests?

**Comment by dsouzai** [comment]
> Can/has Devin's test case be(en) added to our tests?

In theory I suppose it can, but because it involves compiling (using a platforms specific native compiler) the native code as a library, it isn't as straightforward as a test that's purely java code. We'd need to get the help of the test team to best determine how to get this test added.

**Pull request merged**